### PR TITLE
Model export extension & related improvements

### PR DESCRIPTION
**Pull request by dsehnal**
Basic extension for exporting structure models in CIF or BinaryCIF format

- Supports exporting transformed structures
- Do not include suffix for "identity assembly operators"
- Do not include assembly-related categories to export if the structure was composed from an assembly
- structAsymMap parsing tweaks
- Support for opening ZIP files with multiple entries

### CODE REVIEW

**Review by arose — COMMENTED** [review]
Thanks David, works quite nicely overall. I played around with it and found some issue. Some are probably only uncovered by this PR not caused by it.

- 3rb8 assembly
    - some issue with the chain mapping of the DNA enties (id 4 & 5)
    - issues with chain id by auth coloring (by label seems fine)
- 3pqr symmetry indices +/- 1
    - some issue with chain id color theme legend: it shows too many entries for auth (label is ok)
- can we add some info category about the exporter and (if possible) what it contains (e.g. assembly X)?
- are we sure it is ok to keep all categories? should we only keep set of categories we know are ok?
    - thinking of all the database and audit ones
- can there be some status update during export? virus assemblies and symmetry cells take some time
    - actually fails for 1m4x (only cif, bcif is fine) for me (RangeError: Invalid string length)
- when applying "Transform Conformation" the secondary structure categories are missing from the exported model (saw this because for models likely from the pdb archive - having a _database_2.database_id = PDB - the secondary structure is not calculated, assuming it is not given on purpose; maybe we should always calculate sstruc, but still odd that it is missing)

later?
- would be nice to have an option that exports the structure as an assembly, i.e. reuse or create _pdbx_struct_* categories
- support multi-model structure export (maybe just/mostly coordinates to make it simpler?)

**Comment by arose** [comment]
@valasatava fyi, in case you want to check this for your alignment tool use case

File: src/mol-theme/color/chain-id.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ function getAsymIdSerialMap(structure: Structure, type: AsymIdType) {
         const offset = (type === 'auth' ? asymIdOffset?.auth : asymIdOffset?.label) || 0;
         let count = 0;
         m.properties.structAsymMap.forEach(({ auth_id }, label_id) => {
-            const asymId = type === 'auth' ? auth_id : label_id;
+            const asymId = type === 'auth' && auth_id ? auth_id : label_id;
```
[the remark this answers is not in the document]
Comment (dsehnal) [reply src/mol-theme/color/chain-id.ts:65]: Yeah, this was an ugly workaround for then ``auth_id`` is empty (which should not really happen and is prolly a bug somewhere else in the code).

[the remark this answers is not in the document]
Comment (arose) [reply src/mol-theme/color/chain-id.ts:65]: Had a look at this and the issue is that `auth_id` is not set in `getStructAsymMap`. Could you set it there from the `molstar_atom_site_operator_mapping` category? We should probably not use `struct_asym` at all in `getStructAsymMap` as it doesn't include `auth_id`. Can't think of a reason from the top of my head why we would.

[the remark this answers is not in the document]
Comment (dsehnal) [reply src/mol-theme/color/chain-id.ts:65]: Ok, lets drop the ``struct_asym`` support and I will pass ``structAsymMap`` to the re-mapping step too.

> Can't think of a reason from the top of my head why we would.

You tell me, you wrote that part of the code :)

[the remark this answers is not in the document]
Comment (dsehnal) [reply src/mol-theme/color/chain-id.ts]: This should be addressed by 4871f15.

**Comment by dsehnal** [comment]
> * 3rb8 assembly
>       
>       * some issue with the chain mapping of the DNA enties (id 4 & 5)
>       * issues with chain id by auth coloring (by label seems fine)
> 
>     * 3pqr symmetry indices +/- 1
>       
>       * some issue with chain id color theme legend: it shows too many entries for auth (label is ok)

These 2 should be fixed with ``structAsymMap`` being generated differently.

> 
>     * can we add some info category about the exporter and (if possible) what it contains (e.g. assembly X)?

Let's keep this one for later.

> 
>     * are we sure it is ok to keep all categories? should we only keep set of categories we know are ok?
>       
>       * thinking of all the database and audit ones

So this is mostly to support the extra AlphaFold categories "out of the box". Perhaps in a separate PR we could have "exclude category list"?

> 
>     * can there be some status update during export? virus assemblies and symmetry cells take some time
>       
>       * actually fails for 1m4x (only cif, bcif is fine) for me (RangeError: Invalid string length)

Added static overlay. Partial progress would be much harder since the CIF encoder doesn't support the ``Task`` interface.

> 
>     * when applying "Transform Conformation" the secondary structure categories are missing from the exported model (saw this because for models likely from the pdb archive - having a _database_2.database_id = PDB - the secondary structure is not calculated, assuming it is not given on purpose; maybe we should always calculate sstruc, but still odd that it is missing)

Could not reproduce this one (tested superposing 2 structures and exporting them).

> later?
> 
>     * would be nice to have an option that exports the structure as an assembly, i.e. reuse or create _pdbx_struct_* categories

Yeah, but I think this is not the main use case (main use case would be superposed original models & alpha fold structures)

>     * support multi-model structure export (maybe just/mostly coordinates to make it simpler?)

Yeah, doing just atom_site export here seems like the best initial implementation.

File: src/mol-theme/color/chain-id.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (arose) [reply src/mol-theme/color/chain-id.ts]: thanks

File: src/extensions/model-export/export.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                : `${s.model.entryId}.${format}`;
+            entryMap.set(s.model.entryId, (entryMap.get(s.model.entryId) ?? 0) + 1);
+
+            await ctx.update({ message: `Exporting ${s.model.entryId}...`, isIndeterminate: true, canAbort: false });
+            if (s.elementCount > 100000) {
+                // Give UI chance to update, only needed for larger structures.
+                await new Promise(res => setTimeout(res, 50));
+            }
+
+            files.push([name, to_mmCIF(s.model.entryId, s, format === 'bcif', { copyAllCategories: true })]);
```
Comment (arose) [inline src/extensions/model-export/export.ts:53]: maybe catch `RangeError: Invalid string length` and suggest using bcif?

**Review by arose — APPROVED** [review]
Code lgtm. Tried 1rb8, 1m4x (16M atoms), 3sn6, 1sfi (PRD). MOdels with transformed coordinates don't seem to be a problem (anymore). At least superposing 3pqr and 3sn6, then exporting and loading them works fine for me.

**Pull request merged**